explorer/interpreter/trace_stream.h

```diff
@@ -0,0 +1,83 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_INTERPRETER_TRACE_STREAM_H_
+#define CARBON_EXPLORER_INTERPRETER_TRACE_STREAM_H_
+
+#include <optional>
+#include <string>
+
+#include "common/check.h"
+#include "common/ostream.h"
+#include "explorer/common/nonnull.h"
+#include "explorer/common/source_location.h"
+
+namespace Carbon {
+
+// Encapsulates the trace stream so that we can cleanly disable tracing while
+// the prelude is being processed. The prelude is expected to take a
+// disproprotionate amount of time to log, so we try to avoid it.
+//
+// TODO: While the prelude is combined with the provided program as a single
+// AST, we detect the boundary where declarations stop being prelude
+// declarations. When the prelude is fully treated as a separate file, we may be
+// able to take a different approach where the caller explicitly toggles tracing
+// when switching file contexts.
+class TraceStream {
+ public:
+  explicit TraceStream(std::string_view prelude_filename)
+      : prelude_filename_(prelude_filename) {}
+
+  // Returns true if tracing is currently enabled.
+  auto is_enabled() const -> bool {
+    return stream_.has_value() && !skipping_prelude_;
+  }
+
+  // If the prelude is currently being skipped and source_loc is a non-prelude
+  // file location, this will mark the skip as done.
+  auto update_skipping_prelude(SourceLocation source_loc) -> void {
+    if (skipping_prelude_ && !source_loc.filename().empty() &&
```

Comment:
It makes me a little uneasy to be hardcoding here that we will always visit all prelude locations before all non-prelude locations. I wonder if it would be better to separately track a `skipping_prelude_` flag (set by `set_skipping_prelude`) and an `in_prelude_` flag (set by this function), and only skip tracing if both flags are set.

I'd also be inclined to rename this function to something a bit more general like `set_trace_location`, but only very slightly.

Reply:
Per some side discussion about this:

- Moved logic to be local in type_checker
- Replaced skipping_prelude with in_prelude
- Changed the AST to include num_prelude_declarations so that this class no longer tracks with the filename, and eliminating update_skipping_prelude